src/hugr.rs

```diff
@@ -202,9 +202,22 @@ impl Hugr {
         Ok(extension_closure)
     }
 
-    /// TODO: Write this
-    fn instantiate_extensions(&mut self, _solution: ExtensionSolution) {
-        //todo!()
+    /// Add extension requirement information to the hugr in place.
+    fn instantiate_extensions(&mut self, solution: ExtensionSolution) {
+        // We only care about inferred _input_ extensions, because `NodeType`
+        // uses those to infer the output extensions
+        for ((node, _), input_extensions) in solution
+            .iter()
+            .filter(|((_, dir), _)| *dir == Direction::Incoming)
+        {
+            let nodetype = self.op_types.try_get_mut(node.index).unwrap();
+            match nodetype.signature() {
+                None => nodetype.input_extensions = Some(input_extensions.clone()),
+                Some(existing_ext_reqs) => {
```

Comment:
This confused me - let's rename `existing_ext_reqs`. Maybe `sig` or `sig_with_reqs` or something like that?

Reply:
And, `nodetype.input_extensions` still exists. Yet here we neither assign to it, nor assert about it, nor anything else. Should it be equal to the input_extensions of the signature, something like that? This `match` would make a lot more sense if we dealt with (e.g. `debug_assert`ed about) `nodetype.input_extensions` in both cases, rather than just one...

Reply:
`nodetype.input_extensions` should be the same as `existing_ext_reqs.input_extensions`, but I'll refactor to make it more clear

Reply:
I've pushed `cc04a84` which addresses this

Reply:
Yes, this is clearer, thanks :) (I see now what `signature()` does and yes equivalent, but now we don't need to know that)